---
title: "ChIPseq Peak Overlaps"
author: "Tamas Schauer"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ChIPseq Peak Overlaps}
  \usepackage[utf8]{inputenc}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

* ChIP-seq data using H2Av antibody in Drosophila
* 2 replicate datasets
* Peaks were called by Homer (not shown here)

## Setup Annotation

```{r message=FALSE, warning=FALSE,results=FALSE}
library(HelpersforChIPSeq)

### load gene annotation
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)

txdb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
my_genes <- genes(txdb)

my_promoters <- promoters(my_genes, upstream = 1000, downstream = 1000)
my_promoters <- reduce(my_promoters, ignore.strand = TRUE)
my_promoters <- my_promoters[seqnames(my_promoters) %in% "chr2L"]
```

```{r}
my_promoters
```


## Load Data


```{r message=FALSE, warning=FALSE,results=FALSE, eval=TRUE}
### import bed as granges object
data_dir <- system.file("extdata/", package = "HelpersforChIPSeq")
bed_files <- file.path(data_dir, list.files(path = data_dir, pattern = ".bed$"))

ranges_rep1 <- rtracklayer::import(bed_files[1])
ranges_rep2 <- rtracklayer::import(bed_files[2])
```

```{r}
ranges_rep1
```

```{r}
ranges_rep2
```

## Overlap Counting Rules

* define a pooled set of ranges
* overlap: subset of pool overlapping both ranges
* no overlap: subset of pool overlapping only one of the ranges

```{r message=FALSE, warning=FALSE,results=FALSE, echo=FALSE, fig.align='center', fig.height=5, fig.width=9, dpi=200, out.width='75%'}
cbPalette <- c("#E69F00", "#56B4E9", "#999999", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

my_region <- makeGRangesFromDataFrame(data.frame(chr = "chr2L", 
                                                 start = 10714000, 
                                                 end = 10748000))

ranges_rep1_region <- subsetByOverlaps(ranges_rep1, my_region)
ranges_rep2_region <- subsetByOverlaps(ranges_rep2, my_region)

my_granges_names = c("ranges_rep1_region","ranges_rep2_region")

my_pooled_ranges <- GenomicRanges::GRanges()

for(i in seq_along(my_granges_names)){
        my_pooled_ranges <- append(my_pooled_ranges, get(my_granges_names[i]))
}

my_pooled_ranges <- reduce(my_pooled_ranges,
                           min.gapwidth = 0L, 
                           ignore.strand = TRUE)
my_pooled_ranges$range_id <- paste("range", seq_along(my_pooled_ranges), sep="_")

for(i in seq_along(my_granges_names)){
        
        my_overlaps <- !(is.na(findOverlaps(my_pooled_ranges, get(my_granges_names[i]), select="arbitrary")))
        mcols(my_pooled_ranges) <- cbind(mcols(my_pooled_ranges), my_overlaps)
        colnames(mcols(my_pooled_ranges))[i+1] <- my_granges_names[i]
}

par(mfrow=c(6,1),mar = c(1,3,1,3))


plotRanges(ranges_rep1_region, 
           my_region = my_region, 
           my_title = "ranges #1 (n=6)", my_color = cbPalette[1])

plotRanges(ranges_rep2_region,  
           my_region = my_region, 
           my_title = "ranges #2 (n=4)", my_color = cbPalette[2])

plotRanges(my_pooled_ranges,  my_region = my_region,  
           my_title = "pool (n=6)", my_color = cbPalette[3])

plotRanges(my_pooled_ranges[my_pooled_ranges$ranges_rep1_region & my_pooled_ranges$ranges_rep2_region],
           my_region = my_region, 
           my_title = "overlap (n=3)", my_color = cbPalette[7])

plotRanges(my_pooled_ranges[my_pooled_ranges$ranges_rep1_region & !(my_pooled_ranges$ranges_rep2_region)], 
           my_region = my_region, 
           my_title = "no overlap #1 (n=2)", my_color = cbPalette[1])

plotRanges(my_pooled_ranges[my_pooled_ranges$ranges_rep2_region & !(my_pooled_ranges$ranges_rep1_region)], 
           my_region = my_region, 
           my_title = "no overlap #2 (n=1)", my_color = cbPalette[2])

par(fig=c(0,1,0,1), mar=c(0,0,0,0), oma=c(0,0,0,0), new=TRUE)
plot.new()
abline(h = c(4/6,3/6), lty=2)
```



* See details in this [tutorial](https://htmlpreview.github.io/?https://github.com/tschauer/HelpersforChIPSeq/blob/master/doc/ChIPseq_Peak_Overlaps_Tutorial.html)

## Venn diagrams

#### Example

```{r message=FALSE, warning=FALSE,results=FALSE, echo=FALSE, fig.align='center', fig.height=6, fig.width=6, dpi=200, out.width='75%'}
plotVennOverlaps(my_granges_names = c("ranges_rep1_region","ranges_rep2_region"),
                 my_labels = c("rep #1", "rep #2"),
                 my_venn_colors = cbPalette)
```

#### Real data

```{r message=FALSE, warning=FALSE,results=FALSE, fig.align='center', fig.height=6, fig.width=6, dpi=200, out.width='75%'}
cbPalette <- c("#E69F00", "#56B4E9", "#CC79A7", "#009E73", "#F0E442", "#0072B2", "#D55E00")

# 2-way comparison
plotVennOverlaps(my_granges_names = c("ranges_rep1","ranges_rep2"),
                 my_labels = c("rep #1", "rep #2"),
                 my_venn_colors = cbPalette)
```


```{r message=FALSE, warning=FALSE,results=FALSE, fig.align='center', fig.height=8, fig.width=8, dpi=200, out.width='75%'}
# 3-way comparison
plotVennOverlaps(my_granges_names = c("ranges_rep1","ranges_rep2", "my_promoters"),
                 my_labels = c("rep #1", "rep #2", "promoters"),
                 my_venn_colors = cbPalette)
```


```{r message=FALSE, warning=FALSE,results=FALSE, fig.align='center', fig.height=8, fig.width=8, dpi=200, out.width='75%'}
# shift ranges randomly left and right
set.seed(1)
ranges_shifted <- shift(ranges_rep2, shift = sample(-1e5:1e5, size = length(ranges_rep2)))

my_granges_names = c("ranges_rep1","ranges_rep2", "my_promoters", "ranges_shifted")

# 4-way comparison
plotVennOverlaps(my_granges_names = my_granges_names,
                 my_labels = gsub(".*_", "", my_granges_names),
                 my_venn_colors = cbPalette)
```




```{r}
sessionInfo()
```

