---
title: "ChIPseq Composite Plot"
author: "Tamas Schauer"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ChIPseq Composite Plot}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

* ChIP-seq data using H2Av antibody in Drosophila
* 2 replicates
* 2 different normalization methods for fun

## Setup Annotation

```{r message=FALSE, warning=FALSE,results=FALSE}
library(HelpersforChIPSeq)

### include only chr2L for this demo 
my_chromosomes <- c("chr2L")


### load gene annotation
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)

txdb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
my_genes <- genes(txdb)

sites.TSS <- data.frame(chr = seqnames(my_genes),
                        center = ifelse(strand(my_genes) == "+", start(my_genes), end(my_genes)),
                        strand = strand(my_genes))
rownames(sites.TSS) <- my_genes$gene_id

sites.TSS <- sites.TSS[sites.TSS$chr %in% my_chromosomes,]
```

```{r}
head(sites.TSS)
```


## Load Example Data

```{r message=FALSE, warning=FALSE,results=FALSE, echo=FALSE}
data_dir <- system.file("extdata/", package = "HelpersforChIPSeq")
```


```{r message=FALSE, warning=FALSE,results=FALSE, eval=FALSE}
### convert bedgraph to coverage vector
data_dir <- system.file("extdata/", package = "HelpersforChIPSeq")
bedgraph_files <- file.path(data_dir, list.files(path = data_dir, pattern = ".bedgraph.gz$"))


parallel::mclapply(seq_along(bedgraph_files), mc.cores = 4, FUN = function(i){

        my_name <- paste("coverage.",gsub(".INPnorm.*bedgraph.gz|.*\\/","",bedgraph_files[i]), sep="")

        my_bedgraph <- rtracklayer::import(bedgraph_files[i])

        seqlevelsStyle(my_bedgraph) <- "UCSC"
        my_bedgraph <- keepSeqlevels(my_bedgraph, my_chromosomes, pruning.mode = "coarse")

        my_cov <- coverage(my_bedgraph, weight = "score")

        assign(my_name, my_cov)
        save(list = my_name, file = paste(data_dir, my_name, ".rda", sep=""))
})

```

```{r}
### load coverage rda files
coverage_files <- file.path(data_dir,list.files(path = data_dir, pattern = "^coverage..*rda"))

for(i in seq_along(coverage_files)){
        
        load(coverage_files[i])
}

my_covs <- ls(pattern = "^coverage\\.")
my_covs

print(get(my_covs[1]))
```



## Site Centered Coverage Matrix

```{r message=FALSE, warning=FALSE,results=FALSE, echo=FALSE}
library(tsTools)

### center coverage at TSSs
my_site_name <- "sites.TSS"
my_window_size <- 10000
my_binning <- 10
```


```{r message=FALSE, warning=FALSE,results=FALSE, eval=FALSE}
library(tsTools)

### center coverage at TSSs
my_site_name <- "sites.TSS"
my_window_size <- 4000
my_binning <- 10

parallel::mclapply(seq_along(my_covs), mc.cores = 4, FUN = function(i){
        
        my_name <-  paste(gsub("sites","mat",my_site_name), gsub("coverage.","", my_covs[i]), sep=".")
        
        my_cov <- get(my_covs[i])
        
        my_mat <- coverageWindowsCenteredStranded(centers = get(my_site_name),
                                                  window.size = my_window_size,
                                                  coverage = my_cov)
        
        my_mat <- binMeansMatrix(my_mat, my_binning)
        
        assign(my_name, my_mat)
        save(list = my_name, file = paste(data_dir, my_name, ".rda", sep=""))
})
```

```{r}
my_mats <- file.path(data_dir ,list.files(path = data_dir, pattern = "^mat"))

for(i in seq_along(my_mats)){
        
        load(my_mats[i])
}

my_mats <- ls(pattern = "^mat")
my_mats
       
dim(get(my_mats[1]))

get(my_mats[1])[1:5,1:10] 
```

```{r message=FALSE}
### filter out TSSs with "flat" signal
library(matrixStats)

my_filter <- rowSds(averageMats(my_mats)) > 0.25
my_order <- order(rowMeans(averageMats(my_mats)[my_filter,]))


for(j in seq_along(my_mats)){

        my_mat_name <- my_mats[j]
        my_mat <- get(my_mat_name)

        my_mat <- my_mat[my_filter,]
        my_mat <- my_mat[my_order,]

        assign(my_mat_name, my_mat)
}

# check whether row names are identical in all matrices
all(sapply(seq_along(my_mats), function(i){(identical(rownames(get(my_mats[length(my_mats)])), rownames(get(my_mats[i]))))}))

dim(get(my_mats[1]))
```


## Composite Plot

```{r message=FALSE, warning=FALSE,results=FALSE, fig.align='center', fig.height=7, fig.width=7, dpi=200, out.width='75%'}
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
my_ylims <- list(c(1,2),c(1,4),c(1,2),c(1,4))

par(mfrow=c(2,2), mar = c(4,4,2,2), oma = c(1,1,1,1), mgp = c(2,1,0))

for(i in seq_along(my_mats)){
        
        my_mat <- my_mats[i]
        
        plotComposite(my_sample_mats = my_mat, 
                      my_sub_range = 1:ncol(get(my_mat)), 
                      ylims =  my_ylims[[i]], 
                      my_binning = my_binning, 
                      my_colors_composite = cbPalette[i], 
                      my_title = gsub("mat.TSS.","",my_mat),
                      add_range = NULL, 
                      smoother = 1, 
                      line_lwd = 2)  
}


```
   


```{r message=FALSE, warning=FALSE,results=FALSE, fig.align='center', fig.height=7, fig.width=7, dpi=200, out.width='75%'}
par(mfrow=c(2,2), mar = c(4,4,2,2), oma = c(1,1,1,1), mgp = c(2,1,0))

plotComposite(my_sample_mats = my_mats, 
              my_sub_range = (ncol(get(my_mats[1]))*1/4):(ncol(get(my_mats[1]))*3/4), 
              ylims =  c(1,4), 
              my_binning = my_binning, 
              my_colors_composite = cbPalette, 
              my_title = "Comparison",
              add_range = NULL, 
              smoother = 1, 
              line_lwd = 2)  

my_mats_names <- gsub("mat.TSS.","",my_mats)

plotLegendtoPosition(conditions = factor(my_mats_names, levels = unique(my_mats_names)),
                     legend_colors = cbPalette, 
                     horizontal = F,
                     position = c(0.55, 0.85))
```


```{r message=FALSE, warning=FALSE,results=FALSE, fig.align='center', fig.height=7, fig.width=7, dpi=200, out.width='75%'}
par(mfrow=c(2,2), mar = c(4,4,2,2), oma = c(1,1,1,1), mgp = c(2,1,0))

my_mat <- my_mats[4]
my_ylims <- list(c(0,7),c(0,7),c(1.5,4),c(1.5,4))
range_types <- c("SD","IQR","SEM", "CI")

for(i in seq_along(range_types)){
        
        range_type <- range_types[i]
        
        plotComposite(my_sample_mats = my_mat, 
                      my_sub_range = 1:ncol(get(my_mat)), 
                      ylims = my_ylims[[i]], 
                      my_binning = my_binning, 
                      my_colors_composite = cbPalette[7], 
                      my_title = range_type,
                      add_range = range_type, 
                      smoother = 1, 
                      line_lwd = 2)  
}


```

```{r}
sessionInfo()
```



